File: src/main/java/org/bubenheimer/android/preference/DurationPickerPreference.java
/*
 * Copyright (c) 2015-2017 Uli Bubenheimer.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.bubenheimer.android.preference;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.v7.preference.DialogPreference;
import android.support.v7.preference.PreferenceDialogFragmentCompat;
import android.util.AttributeSet;

import org.bubenheimer.android.ui.StyleUtils;
import org.bubenheimer.android.util.R;

import java.util.Locale;

@SuppressWarnings("unused")
public final class DurationPickerPreference extends DialogPreference implements DialogSupporter {
    //Default duration: 2 hours
    static final int DEFAULT_VALUE = 120;

    private int value = DEFAULT_VALUE;

    public DurationPickerPreference(final Context context, final AttributeSet attrs,
                                  final int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    public DurationPickerPreference(final Context context, final AttributeSet attrs,
                                  final int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public DurationPickerPreference(final Context context, final AttributeSet attrs) {
        this(context, attrs, StyleUtils.getAttr(
                context, R.attr.durationPickerPreferenceStyle,
                R.attr.durationPickerPreferenceStyle));
    }

    public DurationPickerPreference(final Context context) {
        this(context, null);
    }

    public int getValue() {
        return value;
    }

    public void setValue(final int value) {
        this.value = value;
        persistInt(value);
        notifyChanged();
    }

    static int getDays(final int minutes) {
        return minutes / (60 * 24);
    }

    static int getHours(final int minutes) {
        return minutes % (60 * 24) / 60;
    }

    static int getMinutes(final int minutes) {
        return minutes % 60;
    }

    @Override
    protected Object onGetDefaultValue(final TypedArray a, final int index) {
        return a.getInteger(index, DEFAULT_VALUE);
    }

    @Override
    protected void onSetInitialValue(final boolean restorePersistedValue,
                                     final Object defaultValue) {
        setValue(restorePersistedValue ? getPersistedInt(value) : (int) defaultValue);
    }

    @Override
    public CharSequence getSummary() {
        final CharSequence summary = super.getSummary();
        if (summary == null) {
            if (value == 0) {
                return "Unlimited";
            } else {
                return String.format((Locale) null, "%dd %dh %dm",
                        getDays(value), getHours(value), getMinutes(value));
            }
        } else {
            return summary;
        }
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        final Parcelable superState = super.onSaveInstanceState();
        // Check whether this Preference is persistent (continually saved)
        if (isPersistent()) {
            // No need to save instance state since it's persistent, use superclass state
            return superState;
        }

        final SavedState myState = new SavedState(superState);
        myState.value = value;
        return myState;
    }

    @Override
    protected void onRestoreInstanceState(final Parcelable state) {
        // Check whether we saved the state in onSaveInstanceState
        if (state == null || !state.getClass().equals(SavedState.class)) {
            // Didn't save the state, so call superclass
            super.onRestoreInstanceState(state);
            return;
        }

        // Cast state to custom BaseSavedState and pass to superclass
        final SavedState myState = (SavedState) state;
        super.onRestoreInstanceState(myState.getSuperState());

        // Set this Preference's widget to reflect the restored state; the widget better exist
        setValue(myState.value);
    }

    @Override
    public PreferenceDialogFragmentCompat newDialog() {
        return DurationPickerPreferenceDialogFragment.newInstance(getKey());
    }

    private static final class SavedState extends BaseSavedState {
        int value;

        SavedState(final Parcelable superState) {
            super(superState);
        }

        SavedState(final Parcel source) {
            super(source);
            value = source.readInt();
        }

        @Override
        public void writeToParcel(final Parcel dest, final int flags) {
            super.writeToParcel(dest, flags);
            dest.writeInt(value);
        }

        public static final Creator<SavedState> CREATOR =
                new Creator<SavedState>() {
                    public SavedState createFromParcel(final Parcel in) {
                        return new SavedState(in);
                    }

                    public SavedState[] newArray(final int size) {
                        return new SavedState[size];
                    }
                };
    }
}

File: src/main/java/org/bubenheimer/android/preference/DurationPickerPreference.kt
/*
 * Copyright (c) 2015-2019 Uli Bubenheimer
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package org.bubenheimer.android.preference

import android.content.Context
import android.content.res.TypedArray
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import androidx.preference.DialogPreference
import androidx.preference.PreferenceDialogFragmentCompat
import org.bubenheimer.android.util.R

public class DurationPickerPreference @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = R.attr.durationPickerPreferenceStyle,
    defStyleRes: Int = 0
) : DialogPreference(context, attrs, defStyleAttr, defStyleRes), DialogSupporter {
    internal companion object {
        //Default duration: 2 hours
        internal const val DEFAULT_VALUE = 120

        fun getDays(minutes: Int): Int {
            return minutes / (60 * 24)
        }

        fun getHours(minutes: Int): Int {
            return minutes % (60 * 24) / 60
        }

        fun getMinutes(minutes: Int): Int {
            return minutes % 60
        }
    }

    public var value: Int = DEFAULT_VALUE
        set(value) {
            field = value
            persistInt(value)
            notifyChanged()
        }

    public override fun onGetDefaultValue(a: TypedArray, index: Int): Any {
        return a.getInteger(index, DEFAULT_VALUE)
    }

    public override fun onSetInitialValue(defaultValue: Any?) {
        value = getPersistedInt(defaultValue as Int? ?: value)
    }

    public override fun getSummary(): CharSequence =
        super.getSummary() ?: (if (value == 0) "Unlimited"
        else String.format("%dd %dh %dm", getDays(value), getHours(value), getMinutes(value)))

    override fun onSaveInstanceState(): Parcelable? = super.onSaveInstanceState().let {
        // No need to save instance state when preference is persistent (continuously saved),
        // use super class state
        if (isPersistent) it
        else SavedState(it ?: BaseSavedState.EMPTY_STATE).also { value = this.value }
    }

    override fun onRestoreInstanceState(state: Parcelable?): Unit =
        // Only restore state ourselves when we actually saved some
        state?.let { it as? SavedState }?.let {
            super.onRestoreInstanceState(it.superState)
            value = it.value
        } ?: super.onRestoreInstanceState(state)

    public override fun newDialog(): PreferenceDialogFragmentCompat =
        DurationPickerPreferenceDialogFragment.newInstance(key)

    private class SavedState : BaseSavedState {
        var value = 0

        constructor(superState: Parcelable) : super(superState)
        private constructor(source: Parcel) : super(source) {
            value = source.readInt()
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            super.writeToParcel(dest, flags)
            dest.writeInt(value)
        }

        companion object CREATOR : Parcelable.Creator<SavedState> {
            override fun createFromParcel(parcel: Parcel) = SavedState(parcel)
            override fun newArray(size: Int) = arrayOfNulls<SavedState>(size)
        }
    }
}
